h2/src/main/org/h2/expression/condition/ConditionAndOrN.java

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright 2004-2020 H2 Group. Multiple-Licensed under the MPL 2.0, and the
+ * EPL 1.0 (https://h2database.com/html/license.html). Initial Developer: H2
+ * Group
+ */
+package org.h2.expression.condition;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import org.h2.engine.Session;
+import org.h2.expression.Expression;
+import org.h2.expression.ExpressionVisitor;
+import org.h2.message.DbException;
+import org.h2.table.ColumnResolver;
+import org.h2.table.TableFilter;
+import org.h2.value.Value;
+import org.h2.value.ValueBoolean;
+import org.h2.value.ValueNull;
+
+/**
+ * An 'and' or 'or' condition as in WHERE ID=1 AND NAME=? with N operands.
+ * Mostly useful for optimisation and preventing stack overflow where generated
+ * SQL has tons of conditions.
+ */
+public class ConditionAndOrN extends Condition {
+
+    /**
+     * The AND condition type as in ID=1 AND NAME='Hello'.
+     */
+    public static final int AND = 0;
+
+    /**
+     * The OR condition type as in ID=1 OR NAME='Hello'.
+     */
+    public static final int OR = 1;
+
+    private final int andOrType;
+    /**
+     * Use an ArrayDeque because we primarily insert at the front.
+     */
+    private final List<Expression> expressions;
+
+    /**
+     * Additional conditions for index only.
+     */
+    private List<Expression> added;
+
+    public ConditionAndOrN(int andOrType, Expression expr1, Expression expr2, Expression expr3) {
+        this.andOrType = andOrType;
+        this.expressions = new LinkedList<>();
```

Comment:
Why a `LinkedList` is used here and another constructor stores a passed `ArrayList`? If different implementations really behave better, it should be documented somehow in comments.

Reply:
Good catch, will fix that. Was because at one point in the evolution of the feature I was doing insert at the front.